integration/quota/quota_test.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package quota
+
+import (
+	"context"
+	"fmt"
+	"hash"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/quota/mysql"
+	"github.com/google/trillian/server"
+	"github.com/google/trillian/server/admin"
+	"github.com/google/trillian/server/interceptor"
+	"github.com/google/trillian/storage/testonly"
+	"github.com/google/trillian/testonly/integration"
+	"github.com/google/trillian/trees"
+	"github.com/google/trillian/util"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func TestRateLimiting(t *testing.T) {
+	maxUnsequenced := 20
+
+	adminClient, logClient, closeFn, err := setupLogServer(maxUnsequenced)
+	if err != nil {
+		t.Fatalf("setupLogServer() returned err = %v", err)
+	}
+	defer closeFn()
+
+	ctx := context.Background()
+	tree, err := adminClient.CreateTree(ctx, &trillian.CreateTreeRequest{Tree: testonly.LogTree})
+	if err != nil {
+		t.Fatalf("CreateTree() returned err = %v", err)
+	}
+	hasherFn, err := trees.Hash(tree)
+	if err != nil {
+		t.Fatalf("Hash() returned err = %v", err)
+	}
+	hasher := hasherFn.New()
+	lw := &leafWriter{client: logClient, hash: hasher, treeID: tree.TreeId}
+
+	// Requests where leaves < maxUnsequenced should work
+	for i := 0; i < maxUnsequenced; i++ {
+		if err := lw.QueueLeaf(ctx); err != nil {
+			t.Errorf("QueueLeaf() returned err = %v", err)
+		}
+	}
+
+	// Some point after now requests should start to fail
+	stop := false
+	timeout := time.After(1 * time.Second)
```

Comment:
I ran it a few hundred times and didn't have it flake. The quota system should begin to block requests *soon-ish*, but I didn't want to assume a specific number, in particular because information_schema is an estimate (although it seems to be pretty precise for small numbers). I'm open to suggestions if you have any.